src/cir.rs

```diff
@@ -0,0 +1,439 @@
+use crate::ast::Predicate;
+use crate::context::{Context, Match};
+use crate::interpreter::Execute;
+use crate::lir::{is_operator, LirInstruction, LirLogicalOperators, LirProgram, Translate};
+use crate::semantics::FieldCounter;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct CirProgram {
+    pub(crate) instructions: Vec<CirInstruction>,
+}
+
+impl CirProgram {
+    pub fn new() -> Self {
+        Self {
+            instructions: Vec::new(),
+        }
+    }
+}
+
+impl Default for CirProgram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum CirInstruction {
+    AndIns(AndIns),
+    OrIns(OrIns),
+    NotIns(NotIns),
+}
+
+#[derive(Debug)]
+pub struct AndIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct OrIns {
+    left: CirOperand,
+    right: CirOperand,
+}
+
+#[derive(Debug)]
+pub struct NotIns {
+    right: CirOperand,
+}
+
+#[derive(Debug, Clone)]
+pub enum CirOperand {
+    Index(usize),
+    Predicate(Predicate),
+}
+
+impl CirOperand {
+    pub fn as_predicate(&self) -> Option<&Predicate> {
+        match &self {
+            CirOperand::Index(_index) => None, // never be here, otherwise something wrong
+            CirOperand::Predicate(p) => Some(p),
+        }
+    }
+
+    pub fn as_index(&self) -> Option<usize> {
+        match &self {
+            CirOperand::Index(index) => Some(*index),
+            CirOperand::Predicate(_p) => None, // never be here, otherwise something wrong
+        }
+    }
+}
+
+fn is_index(cir_operand: &CirOperand) -> bool {
+    match cir_operand {
+        CirOperand::Index(_index) => true,
+        CirOperand::Predicate(_p) => false,
+    }
+}
+
+impl Translate for LirProgram {
+    type Output = CirProgram;
+    fn translate(&self) -> Self::Output {
+        let mut cir = CirProgram::new();
+        cir_translate_helper(self, &mut cir);
+        cir
+    }
+}
+
+#[inline]
+fn reduce_translation_stack(
+    cir_instructions: &mut Vec<CirInstruction>,
+    operator_stack: &mut Vec<LirLogicalOperators>,
+    operand_stack: &mut Vec<CirOperand>,
+) {
+    loop {
+        if operator_stack.len() > 0 {
+            match &operator_stack.last().unwrap() {
+                LirLogicalOperators::And => {
+                    if operand_stack.len() >= 2 {
+                        let right = operand_stack.pop().unwrap();
+                        let left = operand_stack.pop().unwrap();
+                        let and_ins = AndIns {
+                            left: left.clone(),
```

Comment:
Seems that we don't need `clone` here, since `left` and `right` are `pop`ped from operand, we can just move their ownership

Reply:
fixed, thanks.